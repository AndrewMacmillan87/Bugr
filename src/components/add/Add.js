import React from "react";

import PropTypes from "prop-types";

import "./add.css";

const Add = ({
    id,
    extraClasses,
    placeholder,
    name,
    onChange,
    revealFunc,
    onClick,
}) => {
    const addClasses = extraClasses
        ? `add-object-container ${extraClasses}`
        : "add-object-container";

    return (
        <div className={addClasses}>
            <input
                className='object-title-input'
                id={id}
                name={name}
                type='text'
                placeholder={placeholder}
                onChange={(e) => onChange(e)}
                onFocus={(e) => (e.target.placeholder = "")}
                onBlur={(e) => (e.target.placeholder = placeholder)}
            />
            <button className='add-project-btn'>
                <i
                    className='create-object fas fa-check-circle'
                    title='Create project'
                    onClick={() => {
                        onClick();
                        revealFunc("");
                    }}></i>
            </button>
            <button className='add-project-btn'>
                <i
                    className='do-not-create-object fas fa-times-circle'
                    onClick={() => revealFunc("")}></i>
            </button>
        </div>
    );
};

Add.propTypes = {
    id: PropTypes.string,
    extraClasses: PropTypes.string,
    placeholder: PropTypes.string,
    name: PropTypes.string,
    onChange: PropTypes.func,
    revealFunc: PropTypes.func,
    onClick: PropTypes.func,
};

export default Add;
