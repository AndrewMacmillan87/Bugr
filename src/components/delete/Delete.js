import React from "react";

import PropTypes from "prop-types";

import "./delete.css";

export const Delete = ({ title, extraClasses, revealFunc, onClick }) => {
    const deleteClasses = extraClasses
        ? `delete-project-container ${extraClasses}`
        : "delete-project-container";

    return (
        <div className={deleteClasses}>
            <p className='delete-project-msg'>Delete project?</p>
            <div className='delete-project-btn-container'>
                <button className='delete-project-btn'>
                    <i
                        className='do-delete fas fa-check-circle'
                        onClick={() => {
                            onClick();
                            revealFunc("");
                        }}></i>
                </button>
                <button className='delete-project-btn'>
                    <i
                        className='do-not-delete fas fa-times-circle'
                        onClick={() => revealFunc("")}></i>
                </button>
            </div>
        </div>
    );
};

Delete.propTypes = {
    title: PropTypes.string,
    extraClasses: PropTypes.string,
    revealFunc: PropTypes.func,
    onClick: PropTypes.func,
};

export default Delete;
