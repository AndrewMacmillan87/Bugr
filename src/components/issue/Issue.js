import React, { useState } from "react";

import StatusBtn from "../btns/StatusBtn";
import StatusSelector from "../status_selector/StatusSelector";

import PropTypes from "prop-types";

import "./issue.css";

export const Issue = ({ id, description, completion, priority, stage }) => {
    const [showPriorities, setPriorities] = useState(false);
    const [showStages, setStages] = useState(false);
    const [showIssueOptions, setIssueOptions] = useState(false);
    const [showDateOptions, setDateOptions] = useState(false);

    const stageId = `stage-${id}`;
    const priorityId = `priority-${id}`;

    return (
        <div className='issue-container'>
            <div className='issue-description-container'>
                <div className='issue-description'>{description}</div>
                <i
                    className='issue-menu fas fa-ellipsis-v'
                    title='Options'
                    onClick={() => setIssueOptions(true)}></i>
                <div className='issue-options-temp-container'>
                    {showIssueOptions && (
                        <div className='issue-options-container'>
                            <i
                                className='remove-menu-options fas fa-times-circle'
                                onClick={() => setIssueOptions(false)}></i>
                            <ul className='issue-options-list'>
                                <li className='issue-option-list-element'>
                                    New issue
                                </li>
                                <li className='issue-option-list-element'>
                                    Edit
                                </li>
                                <li className='issue-option-list-element'>
                                    Delete
                                </li>
                                <li className='issue-option-list-element'>
                                    Notes
                                </li>
                            </ul>
                        </div>
                    )}
                </div>
            </div>
            <div className='issue-state-btn-container'>
                {showStages && (
                    <StatusSelector status='stage' onClick={setStages} />
                )}
                <StatusBtn
                    id={stageId}
                    value={stage}
                    onClick={() => {
                        setStages(true);
                        setPriorities(false);
                    }}
                />
                {showPriorities && (
                    <StatusSelector status='priority' onClick={setPriorities} />
                )}
                <StatusBtn
                    id={priorityId}
                    value={priority}
                    onClick={() => {
                        setPriorities(true);
                        setStages(false);
                    }}
                />
            </div>
            <div className='date-container'>
                <div className='completion-date'>{completion}</div>
                <i
                    className='issue-menu fas fa-ellipsis-v'
                    title='Date options'
                    onClick={() => setDateOptions(true)}></i>
                <div className='date-options-temp-container'>
                    {showDateOptions && (
                        <div className='date-options-container'>
                            <i
                                className='remove-menu-options fas fa-times-circle'
                                onClick={() => setDateOptions(false)}></i>
                            <ul className='date-options-list'>
                                <li className='date-option-list-element'>
                                    Edit
                                </li>
                                <li className='date-option-list-element'>
                                    Delete
                                </li>
                            </ul>
                        </div>
                    )}
                </div>
            </div>
        </div>
    );
};

Issue.propTypes = {
    id: PropTypes.string,
    description: PropTypes.string,
    completion: PropTypes.string,
    priority: PropTypes.string,
    stage: PropTypes.string,
};

export default Issue;
