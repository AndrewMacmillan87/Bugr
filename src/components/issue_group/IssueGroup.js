import React from "react";

import Issue from "../issue/Issue";
import DeletePopUp from "../delete/Delete";
import AddPopUp from "../add/Add";

import PropTypes from "prop-types";

import "./issue_group.css";

export const IssueGroup = ({
    issueGroupId,
    onChangeGroupName,
    currentData,
    setAddGroupId,
    currentAddGroupId,
    setDeleteGroupId,
    currentDeleteGroupId,
    setEditGroupId,
    currentEditGroupId,
    onChange,
    title,
    issues,
    prioritiesId,
    setPrioritiesId,
    stagesId,
    setStagesId,
    issueOptionsId,
    setIssueOptionsId,
    dateOptionsId,
    setDateOptionsId,
    issueDescription,
    setIssueDescription,
    issueCompletion,
    setIssueCompletion,
    updateAppState,
    createGroup,
    deleteGroup,
    editGroupName,
    createIssue,
    deleteIssue,
    editIssue,
    setDate,
    setPriority,
    setStage,
    options,
}) => {
    return (
        <section className='issue-group-container'>
            <div className='issue-group-title-container'>
                <h2 className='issue-group-title'>{title}</h2>
                <div className='group-options-container'>
                    <i
                        className='add-group fas fa-plus-circle'
                        title='Add group'
                        onClick={() => {
                            setAddGroupId(issueGroupId);
                            setDeleteGroupId("");
                            setEditGroupId("");
                        }}></i>
                    <i
                        className='delete-group fas fa-times-circle'
                        title='Delete group'
                        onClick={() => {
                            setDeleteGroupId(issueGroupId);
                            setAddGroupId("");
                            setEditGroupId("");
                        }}></i>
                    <i
                        className='edit-group fas fa-pencil-alt'
                        title='Edit group name'
                        onClick={() => {
                            setEditGroupId(issueGroupId);
                            setDeleteGroupId("");
                            setAddGroupId("");
                        }}></i>
                </div>
                {currentAddGroupId === issueGroupId && (
                    <AddPopUp
                        extraClasses='group-add-position'
                        placeholder='Group name...'
                        name='groupName'
                        onChange={onChange}
                        revealFunc={setAddGroupId}
                        onClick={createGroup}
                        updateAppState={updateAppState}
                    />
                )}
                {currentDeleteGroupId === issueGroupId && (
                    <DeletePopUp
                        title={title}
                        extraClasses='group-delete-position'
                        revealFunc={setDeleteGroupId}
                        onClick={() => {
                            deleteGroup({
                                projectId: currentData.project_id,
                                groupId: currentDeleteGroupId,
                            });
                            updateAppState();
                        }}
                    />
                )}
                {currentEditGroupId === issueGroupId && (
                    <AddPopUp
                        extraClasses='group-add-position'
                        placeholder='New group name...'
                        name='groupName'
                        onChange={onChangeGroupName}
                        revealFunc={setEditGroupId}
                        onClick={editGroupName}
                        updateAppState={updateAppState}
                    />
                )}
            </div>
            <div className='issue-box'>
                {issues &&
                    issues.map((issue) => {
                        return (
                            <Issue
                                key={issue.issue_id}
                                currentData={currentData}
                                issueId={issue.issue_id}
                                description={issue.description}
                                completion={issue.finish_by}
                                priority={issue.priority}
                                stage={issue.stage}
                                prioritiesId={prioritiesId}
                                setPrioritiesId={setPrioritiesId}
                                stagesId={stagesId}
                                setStagesId={setStagesId}
                                issueOptionsId={issueOptionsId}
                                setIssueOptionsId={setIssueOptionsId}
                                dateOptionsId={dateOptionsId}
                                setDateOptionsId={setDateOptionsId}
                                issueDescription={issueDescription}
                                setIssueDescription={setIssueDescription}
                                issueCompletion={issueCompletion}
                                setIssueCompletion={setIssueCompletion}
                                createIssue={createIssue}
                                deleteIssue={deleteIssue}
                                editIssue={editIssue}
                                setDate={setDate}
                                setPriority={setPriority}
                                setStage={setStage}
                                issueGroupId={issueGroupId}
                                updateAppState={updateAppState}
                                options={options}
                            />
                        );
                    })}
            </div>
        </section>
    );
};

IssueGroup.propTypes = {
    issueGroupId: PropTypes.string,
    onChangeGroupName: PropTypes.func,
    currentData: PropTypes.object,
    setAddGroupId: PropTypes.func,
    currentAddGroupId: PropTypes.string,
    setDeleteGroupId: PropTypes.func,
    currentDeleteGroupId: PropTypes.string,
    setEditGroupId: PropTypes.func,
    currentEditGroupId: PropTypes.string,
    onChange: PropTypes.func,
    title: PropTypes.string,
    issues: PropTypes.array,
    groupName: PropTypes.string,
    prioritiesId: PropTypes.string,
    setPrioritiesId: PropTypes.func,
    stagesId: PropTypes.string,
    setStagesId: PropTypes.func,
    issueOptionsId: PropTypes.string,
    setIssueOptionsId: PropTypes.func,
    dateOptionsId: PropTypes.string,
    setDateOptionsId: PropTypes.func,
    issueDescription: PropTypes.string,
    setIssueDescription: PropTypes.func,
    issueCompletion: PropTypes.string,
    setIssueCompletion: PropTypes.func,
    updateAppState: PropTypes.func,
    createGroup: PropTypes.func.isRequired,
    deleteGroup: PropTypes.func.isRequired,
    editGroupName: PropTypes.func.isRequired,
    createIssue: PropTypes.func.isRequired,
    deleteIssue: PropTypes.func.isRequired,
    editIssue: PropTypes.func.isRequired,
    setDate: PropTypes.func.isRequired,
    setPriority: PropTypes.func.isRequired,
    setStage: PropTypes.func.isRequired,
    options: PropTypes.array,
};

export default IssueGroup;
