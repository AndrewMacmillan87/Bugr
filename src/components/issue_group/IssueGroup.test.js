import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

import IssueGroup from "./IssueGroup";
import Issue from "../issue/Issue";
import Delete from "../delete/Delete";
import Add from "../add/Add";

const issueGroups = require("../../test_project_data/project_a.json");

describe("IssueGroup component", () => {
    test("Renders without crashing", () => {
        const div = document.createElement("div");
        ReactDOM.render(<IssueGroup />, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it("Renders correctly", () => {
        const tree = renderer.create(<IssueGroup />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    describe("IssueGroup HTML elements", () => {
        it("should render a section element", () => {
            const wrapper = mount(<IssueGroup />);
            expect(wrapper.find("section").length).toEqual(1);
        });
        it("should render a h2 element", () => {
            const wrapper = mount(<IssueGroup />);
            expect(wrapper.find("h2").length).toEqual(1);
        });

        it("should render 6 i elements", () => {
            const wrapper = mount(<IssueGroup />);
            expect(wrapper.find("i").length).toEqual(6);
        });

        it("should render 6 div elements", () => {
            const wrapper = mount(<IssueGroup />);
            expect(wrapper.find("div").length).toEqual(6);
        });

        it("should render 2 Issue components", () => {
            const wrapper = mount(
                <IssueGroup issues={issueGroups["groups"][0].issues} />
            );
            expect(wrapper.find(Issue).length).toEqual(2);
        });
    });

    describe("renders after click", () => {
        let props;
        beforeEach(() => {
            props = {
                setAddGroupId: jest.fn(),
                setDeleteGroupId: jest.fn(),
            };
        });
        it("should render the Delete component when the delete group button is clicked", () => {
            const wrapper = mount(<IssueGroup {...props} />);
            wrapper
                .find({ className: "delete-group fas fa-times-circle" })
                .simulate("click");
            expect(wrapper.find(Delete).length).toEqual(1);
        });

        it("should render the Add component when the add group button is clicked", () => {
            const wrapper = mount(<IssueGroup {...props} />);
            wrapper
                .find({ className: "add-group fas fa-plus-circle" })
                .simulate("click");
            expect(wrapper.find(Add).length).toEqual(1);
        });
    });
});
