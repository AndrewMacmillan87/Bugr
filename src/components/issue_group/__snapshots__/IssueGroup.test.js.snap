// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssueGroup component Renders correctly 1`] = `
<section
  className="issue-group-container"
>
  <div
    className="issue-group-title-container"
  >
    <h2
      className="issue-group-title"
    />
    <div
      className="group-options-container"
    >
      <i
        className="add-group fas fa-plus-circle"
        onClick={[Function]}
        title="Add"
      />
      <i
        className="delete-group fas fa-times-circle"
        onClick={[Function]}
        title="Delete group"
      />
    </div>
  </div>
  <div
    className="issue-box"
  />
</section>
`;
